main/src/ca/uwaterloo/flix/language/phase/Desugar.scala

```diff
@@ -0,0 +1,583 @@
+    case WeededAst.FormalParam(ident, mod, tpe0, loc) =>
+      val tpe = tpe0.map(visitType)
+      DesugaredAst.FormalParam(ident, mod, tpe, loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.EqualityConstraint]] `econstr0`.
+    */
+  private def visitEqualityConstraint(econstr0: WeededAst.EqualityConstraint): DesugaredAst.EqualityConstraint = econstr0 match {
+    case WeededAst.EqualityConstraint(qname, tpe01, tpe02, loc) =>
+      val tpe1 = visitType(tpe01)
+      val tpe2 = visitType(tpe02)
+      DesugaredAst.EqualityConstraint(qname, tpe1, tpe2, loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.TypeParams]] `tparams0`.
+    */
+  private def visitTypeParams(tparams0: WeededAst.TypeParams): DesugaredAst.TypeParams = tparams0 match {
+    case params: WeededAst.KindedTypeParams => visitKindedTypeParams(params)
+    case WeededAst.TypeParams.Unkinded(tparams1) =>
+      val tparams = tparams1.map(visitTypeParam).collect { case t: DesugaredAst.TypeParam.Unkinded => t }
+      DesugaredAst.TypeParams.Unkinded(tparams)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.Derivations]] `derives0`.
+    */
+  private def visitDerivations(derives0: WeededAst.Derivations): DesugaredAst.Derivations = derives0 match {
+    case WeededAst.Derivations(classes, loc) => DesugaredAst.Derivations(classes, loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.Case]] `case0`.
+    */
+  private def visitCase(case0: WeededAst.Case): DesugaredAst.Case = case0 match {
+    case WeededAst.Case(ident, tpe0, loc) =>
+      val tpe = visitType(tpe0)
+      DesugaredAst.Case(ident, tpe, loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.RestrictableCase]] `case0`.
+    */
+  private def visitRestrictableCase(case0: WeededAst.RestrictableCase): DesugaredAst.RestrictableCase = case0 match {
+    case WeededAst.RestrictableCase(ident, tpe0, loc) =>
+      val tpe = visitType(tpe0)
+      DesugaredAst.RestrictableCase(ident, tpe, loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.Declaration.Op]] `op0`.
+    */
+  private def visitOp(op0: WeededAst.Declaration.Op): DesugaredAst.Declaration.Op = op0 match {
+    case WeededAst.Declaration.Op(doc, ann, mod, ident, fparams0, tpe0, tconstrs0, loc) =>
+      val fparams = visitFormalParams(fparams0)
+      val tpe = visitType(tpe0)
+      val tconstrs = tconstrs0.map(visitTypeConstraint)
+      DesugaredAst.Declaration.Op(doc, ann, mod, ident, fparams, tpe, tconstrs, loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.Kind]] `kind0`.
+    */
+  private def visitKind(kind0: WeededAst.Kind): DesugaredAst.Kind = kind0 match {
+    case WeededAst.Kind.Ambiguous(qname, loc) => DesugaredAst.Kind.Ambiguous(qname, loc)
+    case WeededAst.Kind.Arrow(k1, k2, loc) => DesugaredAst.Kind.Arrow(visitKind(k1), visitKind(k2), loc)
+  }
+
+  /**
+    * Desugars the given [[WeededAst.Expr]] `exp0`.
+    */
+  private def visitExp(exp0: WeededAst.Expr): DesugaredAst.Expr = exp0 match {
```

Comment:
We tend to name every subexpression, e.g. e1, e2 and so forth.

Reply:
its a bit strange, but it saves on horizontal space and it enables debugging with breakpoints.

Reply:
The parameter, or each pattern?

Reply:
Each body of a pattern.

E.g.

e1 = visitExp...
e2 = visitExp...

Reply:
Roger

Reply:
Same thing for `visitType`?